File: Problem 7.py
from __future__ import print_function, division
import math
import sys

# Borrowed from scipy/lib/six.py (sort of)
if sys.version_info.major == 2:
    import __builtin__
    builtins = __builtin__
    xrange = __builtin__.xrange
else:
    import builtins
    xrange = builtins.range


class Eratosthenes(object):
    plist = [2]

    def __init__(self, ub):
        self.ub = ub

    def __call__(self):
        if len(self.plist) > 0:
            for item in self.plist:
                yield item
        for candidate in range(self.plist[-1], max(3, self.ub)):
            # if (candidate & 1 != 0): continue
            fact_ub = math.sqrt(candidate)

            for prime in self.plist:
                if prime > fact_ub:
                    self.plist.append(candidate)
                    yield candidate
                    break
                if candidate % prime == 0:
                    break

if __name__ == '__main__':
    for idx, prime in enumerate(Eratosthenes(1000000000)()):
        if idx == 10001:
            break
    print("The answer to Euler Problem 7 is", prime)

File: Problem 7.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#ifndef P_IDX
#define P_IDX 10001
#endif 
int isprime(int n, int * p_arr);

int isprime(int n, int * p_arr){
    int i = 0;

    // if ((n & 1) == 0)
    // {
    //     return 0;
    // }

    do
    {
        if (n % p_arr[i] == 0)
        {
            return 0;
        }

        i++;

    } while (p_arr[i] != 0 && p_arr[i] <= sqrt(n));

    return 1;
}

int main(int argc, char const *argv[])
{
    int * p_arr = malloc(P_IDX * sizeof(int));
    int candidate = 3;
    int p_idx = 1;

    p_arr[0] = 2;
    for (int i = 1; i < P_IDX; ++i)
    {
        p_arr[i] = 0;
    }
    do
    {
        if (isprime(candidate, p_arr))
        {
            // printf("%i\n", candidate);
            p_arr[p_idx] = candidate;
            p_idx++;
        }
        candidate += 2;
    } while (p_arr[P_IDX-1] == 0);

    printf("The answer to Euler Problem 7 is %i\n", p_arr[P_IDX-1]);

    free(p_arr);
    return 0;

}